win32: Refactor the NSIS installer

k-takata opened the pull request:
After the discussion in #3485, I have created a new NSIS installer.
The appearance is based on @gpwen's work and also uses MUI2, but the rest of the code is based on the original (non-MUI2) NSIS script and this still relies on dosinst.c. So we don't need to maintain duplicated codes.
The items I listed at https://github.com/vim/vim/issues/3485#issuecomment-425577841 are all implemented.

And also:
* Most of the message boxes were removed like gpwen's.
* The command prompt will not be shown. (`nsExec::Exec` is used instead of `ExecWait`.)
* Uninstallation of the old version will be started after selecting the install location. (same as gpwen's)
* The installer now remembers the selections of the user for generating `_vimrc`. (Key remap and mouse behavior.)
* Unicode mode is turned on for future translation. This requires NSIS 3.0.
  (gpwen's translation uses each encoding for each language. This uses UTF-8 (with BOM) for all languages.)

You can download an installer from here:
https://ci.appveyor.com/project/k-takata/vim-win32-installer/build/175/job/jod0d3d3y9fvet3y/artifacts
(No need to care about the CI status. It's just failed to push the result to GitHub releases.)

Edited:
You can also download it from here:
https://github.com/k-takata/vim-win32-installer/releases/tag/v8.1.0443-newinstaller
Newer version is also avaliable in my later comment.

k-takata committed:
installer: Use LogicLib

Currently we use low level syntax of NSIS; labels and jump commands.
LogicLib provides higher level syntax. (E.g. ${If} ... ${EndIf})
This makes the code much easier.

Also removed codes for Windows 9x.

k-takata committed:
nsis: Skip if libgcc_s_sjlj-1.dll doesn't exist

Next release of mlocati/gettext-iconv-windows might remove the
dependency on the DLL.

k-takata committed:
nsis: Respect NLS setting for GvimExt

Don't install gettext and iconv for GvimExt if NLS setting is off.

k-takata committed:
nsis: Register "EstimatedSize"

Show the estimated size in the control panel.

k-takata committed:
nsis: Adjust some settings

* Use 64 MB dictionary for better compression.
* Support HiDPI.

k-takata committed:
nsis: Use nsExec::Exec instead of ExecWait

It doesn't show a noisy command prompt.

k-takata committed:
nsis: Use MUI2

Mostly based on the @gpwen's work in vim-installer-mui2.

However, this still relies on dosinst.c, because of the discussion in
 #3485.  We don't want to maintain duplicated codes in dosinst.c and
nsis script.

Enable the unicode flag for translation.

k-takata committed:
nsis: Use InstallLib for installing DLLs

It is more reliable and easier to read than the current implementation.

k-takata commented:
Screen shots (Installer):

1. Welcome
![newinstaller-01-welcome](https://user-images.githubusercontent.com/840186/46267909-8dc14f80-c572-11e8-9fa4-3b5698dcce8a.png)

2. License
![newinstaller-02-license](https://user-images.githubusercontent.com/840186/46267916-99147b00-c572-11e8-92f7-5f897986e884.png)

3. Choose components
If this is the first time installation:
![newinstaller-03-1-newinstall](https://user-images.githubusercontent.com/840186/46267947-c7925600-c572-11e8-9454-bc8bf5f5a91d.png)
If an older version of Vim has been already installed, another checkbox for uninstalling will be shown:
![newinstaller-03-2-secondinstall](https://user-images.githubusercontent.com/840186/46267980-e85aab80-c572-11e8-9f77-473e153356a9.png)

4. Choose _vimrc setting
![newinstaller-04-vimrc-setting](https://user-images.githubusercontent.com/840186/46267957-cf51fa80-c572-11e8-954b-4814385dfa31.png)

5. Install location
![newinstaller-05-location](https://user-images.githubusercontent.com/840186/46268103-9c5c3680-c573-11e8-8f1c-16ddb934df62.png)

6. Installing
If an older version of Vim has been already installed, the uninstaller will be started before this. If it is uninstalled, then it will back to this screen.
![newinstaller-06-installing](https://user-images.githubusercontent.com/840186/46268115-ab42e900-c573-11e8-93d9-5bc526bbd05c.png)

7. Finish
![newinstaller-07-finish](https://user-images.githubusercontent.com/840186/46268195-168cbb00-c574-11e8-9a93-c5fd81e85251.png)

k-takata commented:
Screen shots (Uninstaller):

1. Start
![newinstaller-08-uninstall-start](https://user-images.githubusercontent.com/840186/46268258-72574400-c574-11e8-8c10-a09e0d4b3cce.png)

2. Choose components
Now I removed the message boxes while uninstalling and this screen is shown instead.
(The last two checkboxes might be better to be off by default? Not sure.)
![newinstaller-09-uninstall-components](https://user-images.githubusercontent.com/840186/46268270-826f2380-c574-11e8-954f-c2811ed6eccd.png)

3. Uninstalling
![newinstaller-10-uninstall-uninstalling](https://user-images.githubusercontent.com/840186/46268279-8f8c1280-c574-11e8-8c3e-5a32e762d686.png)

4. Finish
![newinstaller-11-uninstall-finish](https://user-images.githubusercontent.com/840186/46268286-96b32080-c574-11e8-9f3a-9720d728e36b.png)

scootergrisen commented:
I suggest having the two "default" option in the same style.
Maybe delete the "default" part because if the radio button is selected as default then maybe not nessary to write default because its already selected as default.

Maybe change "mouse behaviour" to "mouse button behavior".
That way you can remove "button" from the following text and have it shorter.
"Right: context menu, left: visual mode"

k-takata commented:
> Maybe delete the "default" part because if the radio button is selected as default then maybe not nessary to write default because its already selected as default.

I don't agree this. As I wrote above, the new installer remember the selection of a user. When the user execute the new installer next time, the selected button might not be the default setting.

k-takata committed:
nsis: Support translation

Import most part from gpwen's work.

k-takata committed:
nsis: Add workarounds for Japanese NSIS messages

Some messages are too long and not shown properly.
Also the access key was missing in a message.
These should be fixed by the NSIS upstream, I think.

k-takata committed:
nsis: Fix some !ifdef conditions

Newly add HAVE_VIS_VIM and make it disabled by default.
VisVim is for VC 6.0.  Does anyone still use it?

k-takata committed:
nsis: Fix installer directory

Fix that the installer didn't check the old version of installed
directory and also $VIM.
Check $VIM first, then check the old directory from the registry, then
fallback to $PROGRAMFILES(X86)\Vim.

k-takata committed:
nsis: Remove old codes

Remove old codes which are no more needed for MUI2.
"XPStyle on" is also removed. This is automatically enabled in MUI2.
No need to use explicitly.

k-takata commented:
Now translations are avaliable.
Also fixed some minor bugs.

k-takata commented:
You can download the updated installer from here:
https://github.com/k-takata/vim-win32-installer/releases/tag/v8.1.0443-newinstaller2

scootergrisen commented:
Cool :smile_cat: 
If the key remapping is always going to be a 2 choice option would it not be better to have it as a check box with the text "Remap a few keys for Windows behavior..."?
Simpler, shorter, less text to read/understand.

k-takata commented:
> If the key remapping is always going to be a 2 choice option would it not be better to have it as a check box with the text "Remap a few keys for Windows behavior..."?

That might be an option. Then we might be able to create another option to select sourcing `defaults.vim` or `vimrc_example.vim`. Currently, dosinst.c accepts the selection only when executed with the interactive mode. It doesn't have a command line option for that.

k-takata commented:
@brammool Thank you for adding this to the todo.txt:

```
Patch to simplify nsis installer. (Ken Takata, 2018 Sep 24, was #3479)
Now included in #3501, using MUI2.

Improve the installer for MS-Windows.  There are a few alternatives:
- mui2 installer improved by Ken Takata (2018 Sep 29)
```

BTW, now I'm thinking to create the both types of installers in our vim-win32-installer nightly builds.
It can make casual users to try the new installer easily. (Of course, we should state that the installer is experimental.)
What do you think of this?

k-takata committed:
nsis: Copy nsis/gvim.nsi to nsis/gvim-orig.nsi

chrisbra commented:
> BTW, now I'm thinking to create the both types of installers in our vim-win32-installer nightly builds.
> It can make casual users to try the new installer easily. (Of course, we should state that the installer is experimental.)
> What do you think of this?

I like that and I think that would get the feature a good test base. In fact, I actually wanted to suggest just that the other day, but forgot about it :(

k-takata committed:
nsis: Add uninstallation check

TODO: Should we run uninstaller again if it seems to be failed?

k-takata committed:
nsis: Support setting Vi compatibility

k-takata commented:
Rebased and changed the _vimrc setting dialogbox.
Now a user can select vi compatibility: `'cp'`, `'nocp'`, sourcing `defaults.vim` or sourcing `vimrc_example.vim`.
![newinstaller-04-vimrc-setting-compat](https://user-images.githubusercontent.com/840186/46448637-d53e1a80-c7c2-11e8-8e4b-c61c495f94a5.png)

k-takata commented:
The updated installer is here:
https://github.com/k-takata/vim-win32-installer/releases/tag/v8.1.0451-newinstaller

k-takata commented:
@chrisbra 
I created a PR for it: vim/vim-win32-installer#74.

scootergrisen commented:
I think it looks to overwhelming for a new user to see all those options/all that text.
It can look alot simpler if use drop downs.
Instead of all those "default" we could have a "Reset" button.
Also the to titles beginning with "Choose..." basically says the same.
![vim](https://user-images.githubusercontent.com/1836590/46451367-2edc2000-c796-11e8-8bfd-7175c3c68c8d.png)

k-takata committed:
nsis: Use DropLists instead of RadioButtons

The _vimrc setting screen was too noisy.
Also adjust wording.

k-takata committed:
nsis: Add "vim_" prefix to registry values

It might be better to add a prefix to avoid conflicts with the system
values.

k-takata commented:
> I think it looks to overwhelming for a new user to see all those options/all that text.

Agree.

> Instead of all those "default" we could have a "Reset" button.

Adding it seems very difficult.

This is the updated screen shot.
![newinstaller-04-vimrc-setting-compat2](https://user-images.githubusercontent.com/840186/46455131-9caf3880-c7e4-11e8-978a-8d45aeee4936.png)

k-takata commented:
The updated installer is here:
https://github.com/k-takata/vim-win32-installer/releases/tag/v8.1.0451-newinstaller2

scootergrisen commented:
You are writing the same things multiple times which does not help the user.
"Choose _vimrv settings" "Choose the settings" "Vi Compatibility" "Compatibility and enhancements" "Vim with all enhancements" "right and left buttons" "Right: ... Left:".

The 3 areas could be as short as "Compatibility", "Keyboard", "Mouse".

At least the "(Default)" is now consistent.

The "Compatibility and enhancements" could have a "newline" at the end so all 3 areas are using the same style.

If someone can remember another software that uses NSIS with a reset button then we could see how they have done it. I can remember any at the moment but will try and look for it.
Or generally see how other software for Windows does it.

k-takata commented:
So, can you propose concrete messages?

```
Right button
left button               [popup menu, starts visual]
```
I don't agree with this part in your screen shot.
It's very hard to under stand that "Right button" corresponds to "popup menu" and "left button" corresponds to "start visual".

scootergrisen commented:
Not sure what the best way would be so its short but not cryptic.
Maybe use "/" to seperate left/right.
I think the right click menu is called context menu and not popup menu.
It would be more logical to have the left button info to the left and the right button info to the right.

    Mouse:
        Button behavior (left/right):
            Visual mode/context menu
            Select mode/context menu
            Visual mode/extend selection

    Mouse:
        Left/right button behavior:
            Visual mode/context menu
            Select mode/context menu
            Visual mode/extend selection

Sometimes in other software i see they go crazy with spaces and big letters in the english source but i dont know if its correct english or they do it to make it easy to read or something.
Maybe some good at english knows more.

    Mouse:
        Button behavior (Left / Right):
            Visual mode / Context menu
            Select mode / Context menu
            Visual mode / Extend selection

Maybe other can give input of what they thing of the examples.

chrisbra commented:
I like the last shown [dialog](https://github.com/vim/vim/pull/3501#issuecomment-426893588). I don't think the Mouse behaviour description should be shortened as mentioned. This looks too confusing. 
I would call those three sections perhaps "Vi Mode" (instead of Vi Compatibility), not sure of a better short name, "Mappings" (instead of Key remapping for Windows behaviour which is too long) and "Mouse" (instead of Mouse behaviour)

chdiza commented:
@chrisbra Your most recent post in this thread got sent to the vim-dev list as being from "Vim Github Bot" instead of as being from Christian Brabant.

chrisbra commented:
OT: @chdiza yes correct. I changed the sender because I thought it looks a bit better. I don't know how to get the actual sender from the message that is created initially so (might have a look later at this)

scootergrisen commented:
The first drop down currently have "(Default)" as the last item.
I suggest perhaps having them all the first item in the each list.

Not sure if "Vi Mode" or "Vi Compatibility" ir correct but i suggest using whatever is using in Vim.
vim --help says something about Vi mode and Ex mode and Improved Ex mode so if that is what its talking about then i suggest using the same text (mode) so its the same.

Also i was thinking if `<C-V>` etc should be `<Ctrl-V>`.
As new Vim user i did not know what `C` is but im guess its Ctrl.
It seems Vim is using CTRL in :help
Maybe it could be written a better way instead of listing all those `<C-*>`.
But i dont know the details.

scootergrisen commented:
Is the installer using the old icons again where white is transparent?
Can you check the icons.

chrisbra commented:
@chdiza (changed it a bit more). Let's see if the result in the vim-dev list is better now.

chrisbra commented:
> @chdiza (changed it a bit more). Let's see if the result in the vim-dev list is better now.

And if it did not forget to save the rewrite script, it should actually work.

(sorry for spamming this issue)

chdiza commented:
> Let's see if the result in the vim-dev list is better now.

Yes, it's back to normal.  The inbox now lists "Christian Brabant" as the sender.

chdiza commented:
> Yes, it's back to normal. The inbox now lists "Christian Brabant" as the sender.

No wait, I'm wrong.  It looks right when GitHub sends it directly to me, but wrong when I'm looking at the inbox of the account that receives the ML.  It now says "chrisbra via Vim Github", instead of what it should say (and used to say), namely your full name.

k-takata commented:
> Is the installer using the old icons again where white is transparent?

You are right. I wasn't aware that the icons were updated recently. Fixed.

k-takata commented:
Okay all, how about this?

![newinstaller-04-vimrc-setting-compat3](https://user-images.githubusercontent.com/840186/46520181-fd5a7600-c8b5-11e8-9ca1-1ee8faaad008.png)

> "Vi Mode" (instead of Vi Compatibility)

The word "mode" reminds me a different thing: normal mode, insert mode, etc.
How about "Vi / Vim behavior"?

> "Mappings" (instead of Key remapping for Windows behaviour which is too long) and "Mouse" (instead of Mouse behaviour)

Agree.

> Also i was thinking if `<C-V>` etc should be `<Ctrl-V>`.

I was afraid that using `<Ctrl-V>` makes the message too long, but using `Ctrl-V` doesn't make it so long.

k-takata commented:
> The first drop down currently have "(Default)" as the last item.
> I suggest perhaps having them all the first item in the each list.

The first drop down is sorted in the order of the enhancements. The last one is most enhanced.
Other lists don't have much meaning in their orders.

k-takata committed:
nsis: Make uninstalling more precise

* Use UninstallLib to delete gettext and iconv DLLs.
* Delete GvimExt{32,64} and pack directories explicitly.
* Delete the vimfiles directory only if it is empty.
  * Also update the messages.
* Delete the top directory non-recursively.

k-takata commented:
Updated the patches to make the uninstaller more precise.

chrisbra commented:
@chdiza okay reverted back to the old behaviour. If anybody feels differently, please open a new ticket, so we don't spam unrelated issues.

chrisbra commented:
@k-takata looks really nice now.

k-takata commented:
Hi all,

The new MUI2 installer is now available at vim-win32-installer:
https://github.com/vim/vim-win32-installer/releases/tag/v8.1.0454
Please try `gvim_8.1.****_x86-mui2.exe`.

k-takata commented:
@brammool I'm now thinking adding support for 64-bit installer, because there are some requests for that. (E.g. vim/vim-win32-installer#73)
But I have a question about the current NSIS script. It requires some executables have the `w32` suffixes. (E.g. `installw32.exe`, `uninstalw32.exe`, ...)
Why they are needed? If we remove them, we can use the same script for both 32- and 64-bit installer.
Otherwise, we should add many !ifdefs (or we should also use the `w32` suffixes for the 64-bit installer.)

k-takata committed:
nsis,dosinst: Support creating 64-bit installer

* Use $PROGRAMFILES64 for the default folder.
* Adjust file names.
* Add " (x64)" to product name for 64-bit.

scootergrisen commented:
Its a little confusing the uninstall dialog just pops up at the beginning of the installation progress.
Could the uninstall stuff perhaps be part of the same dialog so there is not two dialogs.

k-takata commented:
> Could the uninstall stuff perhaps be part of the same dialog so there is not two dialogs.

It's not possible (for now).
Actually, gpwen's MUI2 installer does call the uninstaller silently, but it is because his uninstaller is more precise. It deletes the files only installed by the installer. It doesn't delete `_vimrc` if it is changed by the user.

One possible option is that if the uninstaller is executed silently, do only the first two components: "Unregister Vim" and "Remove Vim Executables/Runtime files". Do not delete vimfiles directory and the root of Vim directory.

k-takata commented:
Added minimal support for creating 64-bit installer.
It still uses the `w32` suffixes even for the 64-bit installer.

k-takata commented:
Bram's comment was not redirected here:
https://groups.google.com/d/msg/vim_dev/TRBJGgCvHik/CEAh1NSGBAAJ

> The "w32" addition is to make sure it doesn't get confused with the
> other install.exe that can be build.  It's possible to use "win" instead
> of "w32" if that is clearer.  But if there actually are two versions of
> the executable, it's probably best to give them different names,
> otherwise it's very easy to end up with the wrong one. 

We have already dropped support for MS-DOS, so the "w32" (or "win") addition might not be so important now.
And yes, we need the 64-bit version of installer.exe to create 64-bit Vim installer.
But I think keep using "w32" is the simplest way. It doesn't require changing a batch file in vim-win32-installer. I have already done a preparation of 64-bit installer in vim/vim-win32-installer#79 in that way.

k-takata commented:
https://github.com/vim/vim-win32-installer/releases/tag/v8.1.0466
64-bit installer is now available at vim-win32-installer. (32-bit build seems failed with some reason.)

k-takata committed:
nsis: Improve some messages

Show a message while waiting the uninstall of old version, or while
executing install.exe or uninstal.exe.

k-takata commented:
Improved a message while waiting uninstall of old version.
![newinstaller-06-installing-wait-uninstall](https://user-images.githubusercontent.com/840186/46591080-20be3480-caf3-11e8-88b8-1e5474afc1e0.png)
Confusion might be reduced.

chrisbra commented:
> https://github.com/vim/vim-win32-installer/releases/tag/v8.1.0466
64-bit installer is now available at vim-win32-installer. (32-bit build seems failed with some reason.)

Rebuild the commit and this time it was successfull

k-takata committed:
nsis: Update Filelist

k-takata commented:
> > Could the uninstall stuff perhaps be part of the same dialog so there is not two dialogs.
> 
> It's not possible (for now).

I'm rethinking this now.
It's not impossible to run the uninstaller in silent mode at the first step of the installing.
But it's not clear which items should be uninstalled and which should be kept.
Someone may want to just update the version of Vim and want to keep the system _vimrc.
Someone may want to reinstall Vim with changing the settings of the system _vimrc.
Kicking the uninstaller in silent mode cannot handle this situation well...

Should we keep the system _vimrc when the uninstaller in silent mode?
Then if someone wants to change the setting of the system _vimrc, he should run the uninstaller manually and next execute the installer. Is it okay?
(gpwen's uninstaller checks if the system _vimrc is changed or not, but it is not easy in the current system.)

I think that the current implementation is better because it is flexible, even if it is somewhat confusing and a bit noisy.
And even so, it is more silent than the original non-MUI2 installer.

scootergrisen commented:
Maybe look at how other software is doing it.
If i click cancel in the uninstall dialog Vim seems to install anyway.

k-takata commented:
> If i click cancel in the uninstall dialog Vim seems to install anyway.

Good point! The behavior is not changed by my patches.
What will happen when a user cancel the uninstaller?
If gvimext{32,64} is in use, it might not be properly updated.
Any other issues? Maybe no.

> Maybe look at how other software is doing it.

Some software ask the user to uninstall the old version before install it.
Some other just overwrite the old version.

scootergrisen commented:
Does both the 32-bit and 64-bit installer use "Program Files (x86)" as default?
I think it should be "Program Files" for 64-bit and "Program Files (x86)" for 32-bit.
More work sorry :cake:

k-takata commented:
> Does both the 32-bit and 64-bit installer use "Program Files (x86)" as default?

No. 64-bit installer use "Program Files" as default, however,if another version of Vim is installed in another directory (e.g. "Program Files (x86)"), it is used as a default.

k-takata commented:
> > > Could the uninstall stuff perhaps be part of the same dialog so there is not two dialogs.
> > 
> > 
> > It's not possible (for now).
> 
> I'm rethinking this now.

It turns out that kicking the uninstaller in silent mode is not reliable in this system.
We cannot know when the uninstaller finished.

So my conclusion is still not changed:

> I think that the current implementation is better because it is flexible, even if it is somewhat confusing and a bit noisy.

k-takata commented:
Added some comments to the system `_vimrc`.
See the latest commit for detail: https://github.com/vim/vim/pull/3501/commits/d1da147a8ea27ed25c45976d44821a34444922ee
E.g.:

```vim
" Vim with all enhancements
source $VIMRUNTIME/vimrc_example.vim

" Remap a few keys for Windows behavior
source $VIMRUNTIME/mswin.vim

" Mouse behavior (the Windows way)
behave mswin

" Use the internal diff if available.
" Otherwise use the special 'diffexpr' for Windows.
if &diffopt !~# 'internal'
  set diffexpr=MyDiff()
endif
```

(Wording might be needed to be adjusted.)

k-takata committed:
nsis: Translate some messages

k-takata commented:
Refactored the commits for easier reviewing.
(Some of them are squashed and some others are edited.)

k-takata commented:
I think I have almost done my work for this.
(If #3525 will be merged before this, I should rebase this upon it, but it might not be a problem.)

If anyone wants to update the translation, please send a PR to the `topic/refactor-installer` branch in https://github.com/k-takata/vim .

scootergrisen commented:
Guess we need some translation instruction on how to translate the installer.
Do i just make a copy of english.nsi?

Are strings like "Typical", "Minimal", "Full" not already translated in the NSIS translations?
They seem so commonly used.

k-takata commented:
> Do i just make a copy of english.nsi?

I think it's the best way for now. The english.nsi has redundant header compared other files, though.
It is based on gpwen's work, some of the messages are not used for now.

> Are strings like "Typical", "Minimal", "Full" not already translated in the NSIS translations?

Yes, they can be translated. See "str_type_typical", "str_type_minimal" and "str_type_full".

scootergrisen commented:
I meant if "Typical", "Minimal", "Full" was already translate in the NSIS translations that comes with NSIS but i guess not.

I dont like that translation strings are wrapping on multiple line so you have:

    "text \[newline][space][space][space][space][space]text"

scootergrisen commented:
Maybe just have each translation on one line like https://github.com/qbittorrent/qBittorrent/blob/master/dist/windows/installer-translations/danish.nsi

k-takata commented:
I just followed the gpwen's way. If you don't like that, blame to him/her.

scootergrisen commented:
I see some leading/trailing spaces i suggest to remove.

```
LangString str_msg_compat_title    ${LANG_ENGLISH} \
    " Vi / Vim behavior "
LangString str_msg_keymap_title   ${LANG_ENGLISH} \
    " Mappings "
LangString str_msg_keymap_default ${LANG_ENGLISH} \
    " Do not remap keys (Default)"
LangString str_msg_keymap_windows ${LANG_ENGLISH} \
    " Remap a few keys"
LangString str_msg_mouse_title   ${LANG_ENGLISH} \
    " Mouse "

```
```

LangString str_msg_compat_title    ${LANG_ENGLISH} \
    "Vi / Vim behavior"
LangString str_msg_keymap_title   ${LANG_ENGLISH} \
    "Mappings"
LangString str_msg_keymap_default ${LANG_ENGLISH} \
    "Do not remap keys (Default)"
LangString str_msg_keymap_windows ${LANG_ENGLISH} \
    "Remap a few keys"
LangString str_msg_mouse_title   ${LANG_ENGLISH} \
    "Mouse"

```

scootergrisen commented:
I see "/" without and with space before/after. Maybe make it consistent whichever is prefered:

    "Remove Vim Executables/Runtime Files"
    " Vi / Vim behavior "

k-takata commented:
I intentionally  added the spaces. Without the spaces, they look too dense.
If your translation doesn't needs the spaces, you can remove them.

scootergrisen commented:
If you want space before/after a string i think its better to add it in the code.
That way translations can focus on translating the text only.

k-takata commented:
I don't agree that.

scootergrisen commented:
There are some strings that starts with "Fail to ".
I think they should be "Failed to ".

scootergrisen committed:
Add danish.nsi

k-takata commented:
The strings were copied from the gpwen's work, but currently not used.

k-takata committed:
nsis: Include danish.nsi

k-takata committed:
nsis: Comment out unused translation messages

They were needed for gpwen's version. Currently, they are not used.

k-takata committed:
nsis: Remove needless spaces

k-takata commented:
I included your Danish translation.

> I see some leading/trailing spaces i suggest to remove.

Oops, some of them should have been removed when I changed the radio buttons to the drop down lists.

k-takata commented:
Rebased on top of 8.1.0472.

k-takata commented:
@brammool 
I think you always get the patch of a PR by adding ".diff" to the end of the URL.
(E.g. https://github.com/vim/vim/pull/3501.diff)
However, I found that this doesn't handle binary files well.
Please use the following zip file for the nsis icon files of this PR.

[nsis-icons.zip](https://github.com/vim/vim/files/2475621/nsis-icons.zip)

k-takata commented:
Or you can use ".patch" instead of ".diff". E.g.

```shell
wget https://github.com/vim/vim/pull/3501.patch
git apply 3501.patch
```

The differences are:
* ".patch" can handle binary files, but ".diff" cannot.
* ".patch" includes all commits and commit messages. (e.g. This PR includes nearly 30 commits.)
 ".diff" just includes the total changes.

k-takata committed:
mui2: Show abort warning

The original installer shows this, but gpwen's didn't show this.
Make the same as the original one.

brammool commented:
Ken Takata wrote:

> @brammool 
> I think you always get the patch of a PR by adding ".diff" to the end of the URL.
> (E.g. https://github.com/vim/vim/pull/3501.diff)
> However, I found that this doesn't handle binary files well.
> Please use the following zip file for the nsis icon files of this PR.
> 
> [nsis-icons.zip](https://github.com/vim/vim/files/2475621/nsis-icons.zip)

Yes, I usually get the .diff file, and that works.  I'll check the icons
once we get to including this.

-- 
hundred-and-one symptoms of being an internet addict:
200. You really believe in the concept of a "paperless" office.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

chrisbra commented:
I just downloaded https://github.com/vim/vim-win32-installer/releases/download/v8.1.0490/gvim_8.1.0490_x64-mui2.exe and installed it on a windows 10 system. 
I really liked the new installer and it worked really well.

k-takata commented:
@chrisbra Thank you for trying!
You might want to send updated German translations? ;-)

chrisbra commented:
@k-takata 
I did not know I am the nsis translator. But here we go: https://gist.github.com/chrisbra/030431f24f6f26ddda559d63e436d67b (hopefully the link finally works after the recent github failure).

2 observations:
 - I did not see the VisVim installation dialog (is this obsolete by now? But it is still uncommented in the language file)
- After a second thought, I think the difference between "vi compatible", "Vim original", "Vim with some enhancements" and "Vim with all enhancements" is not entirely clear. The first ones should be clear to users (&cp setting), but not the last two. Perhaps it is better to mention what file it will source, e.g. 

"Vim with some enhancements (sourcing defaults.vim)" and "Vim with all enhancements (sourcing vimrc_example.vim)"

Hopefully that is not too long.

k-takata committed:
Update German translations

(by @chrisbra)

k-takata commented:
> * I did not see the VisVim installation dialog

I currently disable it, because it is for VC6 or older versions. Is there anyone still use it?

> * After a second thought, I think the difference between "vi compatible", "Vim original", "Vim with some enhancements" and "Vim with all enhancements" is not entirely clear.

Agree. 

> Hopefully that is not too long.

I had to update the layout. Not sure that the droplists should be left aligned or right aligned.

After applying your translation, the description for the key remapping is too long to display:

![newinstaller-04-vimrc-setting-german](https://user-images.githubusercontent.com/840186/47304142-b7950000-d660-11e8-9fa9-fd5865f15369.png)

k-takata committed:
nsis: Adjust some messages and adjust layout for them

* Clarify that loading defaults.vim or vimrc_example.vim.
  (Based on the German translation.)
* Clarify that remap some keys for Windows.

chrisbra commented:
thanks for noticing. Does this [updated patch](https://gist.github.com/chrisbra/030431f24f6f26ddda559d63e436d67b/revisions#diff-29679af159aae6c3cd407caf92622041) fix the issue? It moves some of the description to the title (what is now called Mappings is then called "Mappings to Windows default keys")?

BTW: Is there an easy way to test the translation even when installing on a system with a different locale (or do I have to configure Windows first to use the german system language before trying the installer)?

k-takata commented:
I applied it. It looks okay now.
I also adjusted English messages.

German:
<img alt="newinstaller-04-vimrc-setting-german2" src="https://user-images.githubusercontent.com/840186/47305862-c6ca7c80-d665-11e8-836c-4ca36d734989.png">

English:
<img alt="newinstaller-04-vimrc-setting-english" src="https://user-images.githubusercontent.com/840186/47305879-d21da800-d665-11e8-9ff3-66a56d7d71e7.png">


> BTW: Is there an easy way to test the translation even when installing on a system with a different locale (or do I have to configure Windows first to use the german system language before trying the installer)?

Currently, no. I edited the NSIS script.
Should we show a dialogbox to select the language?

chrisbra commented:
thanks looks good. Can you please finally squash this last change in: https://gist.github.com/chrisbra/030431f24f6f26ddda559d63e436d67b/revisions#diff-29679af159aae6c3cd407caf92622041

> Should we show a dialogbox to select the language?

Yeah, I think this is the easiest way.

k-takata commented:
Updated.

> Yeah, I think this is the easiest way.

Maybe it is useful for debugging, but is it useful for normal users?

chrisbra commented:
> Maybe it is useful for debugging, but is it useful for normal users?

Well, it is probably not used much, but in the rare occasions you want to see translated messages
it is really useful. The alternative is fiddling with Windows to change the system locale, having to download the language pack, restart the system several times before the installer picks up the language setting. 

So a simple selection dialog seems to provide a good compromise if the user decides to install Vim with a different setting and normal users can simply skip the default setting. A command line parameter would work to, but that does not seem to be supported, is it?

brammool commented:
> > * I did not see the VisVim installation dialog
> 
> I currently disable it, because it is for VC6 or older versions. Is
> there anyone still use it?

I suppose someone using MSVC for editing will have a recent version.
We support building Vim with older MSVC compilers, but I don't think it
is very useful to support older IDEs.

Let me know if this is now ready to include.

For reviewers, please let us know if you are missing anything that was
in the old installer.

-- 
Did you hear about the new 3 million dollar West Virginia State Lottery?
The winner gets 3 dollars a year for a million years.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

k-takata committed:
nsis: Add the language selection dialog

The selected language will be stored at `HKCU\Software\Vim\`.

k-takata commented:
I added a language selection dialogbox. The dialogbox will be normally shown only once, and the selected language will be stored at `HKCU\Software\Vim` in the registry. If you delete the key, the dialogbox will be shown again.

k-takata committed:
nsis: Move some functions after the sections

Section IDs might be needed in the functions.

k-takata committed:
nsis: Remember the user's custom selections

Store the selections to the registry.

k-takata commented:
There is a request that the installer should remember the user's custom selection.
Now the installer remember the selected components and select the same components on next installation.

h-east commented:
@k-takata Cool! thanks!! (It's my request)

k-takata commented:
@scootergrisen I have updated some English messages a few days ago.
https://github.com/vim/vim/pull/3501/commits/4ab355812c5b876569e1ede2770c7b0fef46dbcd#diff-ff3c2f200bc53b577622a2fe2f586cd9
If you want to update Danish translation, please send me a PR (or patch).

scootergrisen committed:
Update danish.nsi

scootergrisen commented:
@k-takata danish update https://github.com/k-takata/vim/pull/2

scootergrisen commented:
Instead of "Native Language Support" cant we just say "Translations"?

    LangString str_section_nls          ${LANG_ENGLISH} \
        "Native Language Support"
    LangString str_desc_nls             ${LANG_ENGLISH} \
        "Install files for native language support."

    LangString str_section_nls          ${LANG_ENGLISH} \
        "Translations"
    LangString str_desc_nls             ${LANG_ENGLISH} \
        "Install translations files."

Also i guess "native language" might be incorrect if it installs all languages and not just one.

scootergrisen commented:
Maybe a little confusing the installer file name is called gvim_\*.exe but the installer says vim and control center says vim.
Maybe just call it vim_\*.exe.

scootergrisen commented:
In the "Open with..." context menu it says "Vi Improved - A Text Editor".

Maybe it should just be "Vi Improved" in the context menu so its just the name of the program.

uninstall-gui.exe says "uninstall-gui.exe" in the .exe file description.

k-takata committed:
nsis,dosinst: Make uninstaller silent when installing

* Write "AllowSilent=1" to the registry so that it describe that this
  uninstaller supports silent mode.
* If the installer found "AllowSilent=1", kick the uninstaller with
  the silent mode.
* Delete only the vimfiles directory in the installed directory.
  Not in the $HOME directory.
* Adjust some English messages.

k-takata commented:
I have changed the uninstaller to uninstall silently when executed while installation.

* This checks the registry and if "AllowSilent" is set to 1, the uninstaller will be executed silently.
  If it is not set (which means the uninstaller is old), the uninstaller will be still executed normally.
* The uninstaller is changed not to delete the vimfiles directories in $HOME. Now it deletes only the vimfiles directories in $VIM. (This is for making silent uninstallation safer.)
* English messages are adjusted for above changes.

k-takata commented:
> * English messages are adjusted for above changes.

German or Danish translations might be needed to be updated.

chrisbra commented:
@k-takata is it just `str_msg_uninstalling`?

Then this patch should do it:
```patch
diff --git a/nsis/lang/german.nsi b/nsis/lang/german.nsi
index b188fab33..0e0e1b846 100644
--- a/nsis/lang/german.nsi
+++ b/nsis/lang/german.nsi
@@ -221,7 +221,7 @@ LangString str_msg_rm_exe_fail   ${LANG_GERMAN} \
 #     weil es nicht leer ist!"

 LangString str_msg_uninstalling  ${LANG_GERMAN} \
-    "Uninstalling the old version..."
+    "Deinstalliere alte Version..."

 LangString str_msg_registering   ${LANG_GERMAN} \
     "Registriere..."

k-takata commented:
I also changed `str_desc_rm_vimfiles`, because the uninstaller doesn't delete $HOME/vimfiles.

chrisbra commented:
Okay, adding that one makes the patch:

```patch
diff --git a/nsis/lang/german.nsi b/nsis/lang/german.nsi
index b188fab33..15e47089d 100644
--- a/nsis/lang/german.nsi
+++ b/nsis/lang/german.nsi
@@ -148,7 +148,7 @@ LangString str_desc_rm_exe          ${LANG_GERMAN} \
 LangString str_unsection_vimfiles   ${LANG_GERMAN} \
     "Entferne vimfiles Verzeichnis"
 LangString str_desc_rm_vimfiles     ${LANG_GERMAN} \
-    "Entfernt eigenes vimfiles Verzeichnis, falls es leer ist."
+    "Entfernt das vimfiles Verzeichnis, falls es leer ist."

 LangString str_unsection_rootdir    ${LANG_GERMAN} \
     "Entferne Vim Installationsverzeichnis"
@@ -221,7 +221,7 @@ LangString str_msg_rm_exe_fail   ${LANG_GERMAN} \
 #     weil es nicht leer ist!"

 LangString str_msg_uninstalling  ${LANG_GERMAN} \
-    "Uninstalling the old version..."
+    "Deinstalliere alte Version..."

 LangString str_msg_registering   ${LANG_GERMAN} \
     "Registriere..."
```

(let me know, if you'd like to get the translations as separate PR to your vim fork)

k-takata committed:
nsis: Update German translation

Import https://github.com/vim/vim/pull/3501#issuecomment-435864079

scootergrisen committed:
Update danish.nsi

scootergrisen commented:
I need step by step instructions on how to undate the translations properly.
I'm not sure if i have missing some changes.

chrisbra commented:
@scootergrisen 
check out this pr:

    # (optionally clone the vim repository if you haven't yet)
    git clone https://github.com/vim/vim.git
    cd vim
    # fetch github PR:3501 and checkout to local branch nsis
    git fetch origin pull/3501/head:nsis && git checkout nsis
    cd nsis/lang
    vim danish.nsi
    # check the translations of str_msg_uninstalling and str_desc_rm_vimfiles
    # save and copy paste the diff
    git diff danish.nsi

that should be about it, I think.

scootergrisen commented:
And how would i know i need to use 3501 and check the translations of str_msg_uninstalling and str_desc_rm_vimfiles?

chrisbra commented:
3501 is the id of this PR. the two string variables have been mentioned by me and @k-takata (and I got them from looking at the latest diff from @k-takata)

scootergrisen commented:
I would like step by step instructions that works everytime i and others want to update.
Looking at the latest diff properly works if there have only been one change but what when i return in 6 months and want to update a translation and there have been 100 changes since the last time the translation was updated.

k-takata commented:
https://github.com/vim/vim-win32-installer/releases/tag/v8.1.0513
Starting from this version, the installer can execute the uninstaller silently. (It affects only when uninstalling this version or later, though.)

@scootergrisen This PR has not been merged into the official repository, so the procedure can be changed. Basically, you need to check the changes of `nsis/lang/english.nsi` and apply them to your translation.
If I add a new English message, I may also add them to each translation as is. Because NSIS doesn't fallback to an English message if a corresponding translated message were not found. 
If I modify an English message, I may modify only English.

chrisbra commented:
@scootergrisen 
It's the translation of the installer. I wouldn't expect hundreds of newly to translated strings the next time you look at it. Once it is merged, it is probably pretty stable.

twangboy commented on nsis/gvim.nsi:
```diff
@@ -299,251 +418,627 @@ Section "Vim executables and runtime files"
+FunctionEnd
+
+Function .onInstFailed
+  MessageBox MB_OK|MB_ICONEXCLAMATION "$(str_msg_install_fail)"
```
Please add a default option here for Silent installs `/SD MB_OK`

k-takata replied:
I have added the option, but is it really okay that the installer omits the error message in the silent mode?

twangboy replied:
Optionally you could print out to the console instead...
`$IfSilent` of course

https://stackoverflow.com/questions/3103226/nsis-printing-to-prompt-during-command-line-install

k-takata replied:
I have updated the patch.
`/SD IDOK` is correct, not `MB_OK`.

twangboy replied:
Looks good. I'm glad to finally see some progress on this!

twangboy commented on nsis/gvim.nsi:
```diff
@@ -555,59 +1050,68 @@ Section Uninstall
-	      "Some files in $0 have not been deleted!$\nYou must do it manually."
-	  NoErrorMess:
+	${If} ${Errors}
+	  MessageBox MB_OK|MB_ICONEXCLAMATION $(str_msg_rm_exe_fail)
```
And here...

k-takata committed:
nsis: Don't show any message boxes in the silent mode

k-takata commented:
Rebased on top of 8.1.0535.

k-takata commented:
FYI, these are the recent download counts of the vim-win32-installer releases:

![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0536/total.svg)
![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0536/gvim_8.1.0536_x64-mui2.exe.svg)
![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0536/gvim_8.1.0536_x64.zip.svg)
![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0536/gvim_8.1.0536_x86-mui2.exe.svg)
![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0536/gvim_8.1.0536_x86.exe.svg)
![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0536/gvim_8.1.0536_x86.zip.svg)

(To see above results, use a Web browser.)
At this moment, the download counts are:

v8.1.0536                 | Download counts
--------------------------|----
Total                     | 908
x64 zip                   | 474
x64 mui2 installer        | 243
x86 traditional installer |  98
x86 zip                   |  64
x86 mui2 installer        |  26

The x64 mui2 installer is already downloaded more than twice of x86 traditional installer.
It seems that most people want to use x64 version (even it doesn't have so many advantages to x86).
(As I told in the VimConf Hackathon.)

x64/x86 mui2 installers are downloaded by more than 250 people and we don't have any bug reports for them. I think that that mui2 installer is now quite stable and it is ready to be merged.

k-takata commented:
FYI, these are the download counts before we set up mui2 installers:

![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0451/total.svg)
![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0451/gvim_8.1.0451_x64.zip.svg)
![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0451/gvim_8.1.0451_x86.exe.svg)
![ ](https://img.shields.io/github/downloads/vim/vim-win32-installer/v8.1.0451/gvim_8.1.0451_x86.zip.svg)

v8.1.0451                 | Download counts
--------------------------|----
Total                     | 937
x64 zip                   | 676
x86 traditional installer | 166
x86 zip                   |  72

k-takata committed:
nsis: Support removing private vimfiles directory

Also update the $(str_desc_plugin_home) message for 8.1.0534.
Also changed the messages from "vimfiles directory" to "plugin
directories".

k-takata commented:
I said this is ready to be merged, however, I found some more things to be updated.

* Some messages should be updated to reflect the changes in 8.1.0534.
* Update the uninstaller to be able to remove $HOME/vimfiles directory.
* Some messages are inconsistent between the installer and the uninstaller.
  The installer uses "plugin directories" and the uninstaller uses "vimfiles directory".

chrisbra commented:
so you need updated translations?

k-takata commented:
@chrisbra, @scootergrisen Could you please update the following translations?

* `$(str_desc_plugin_home)`: Updated the message for 8.1.0534. Now this option doesn't create the vimfiles directory in the Vim installation directory.
* `$(str_ungroup_plugin)`, `$(str_desc_rm_plugin)`: The names and the messages were changed.
* `$(str_unsection_plugin_home)`, `$(str_desc_rm_plugin_home)`, `$(str_unsection_plugin_vim)`, `$(str_desc_rm_plugin_vim)`: Newly added.

Ether commenting here or creating a PR should be fine.

chrisbra commented:
Why is `str_ungroup_plugin`  and `str_desc_rm_plugin` in English plural? ("Remove plugin directories") Where does it remove the plugin directories from?

I think this patch should be applied:
```patch
diff --git a/nsis/lang/german.nsi b/nsis/lang/german.nsi
index f0ffffb02..4132bf9bc 100644
--- a/nsis/lang/german.nsi
+++ b/nsis/lang/german.nsi
@@ -116,8 +116,7 @@ LangString str_desc_plugin          ${LANG_GERMAN} \
 LangString str_section_plugin_home  ${LANG_GERMAN} \
     "Privat"
 LangString str_desc_plugin_home     ${LANG_GERMAN} \
-    "Plugin-Verzeichnis im Benutzerverzeichnis (falls es existiert) oder im \
-     Vim-Installationspfad erstellen."
+    "Erstelle Plugin-Verzeichnis im HOME Benutzerverzeichnis."

 LangString str_section_plugin_vim   ${LANG_GERMAN} \
     "Freigegeben"
@@ -146,19 +145,19 @@ LangString str_desc_rm_exe          ${LANG_GERMAN} \
     "Alle Vim-Anwendungen und Laufzeitdateien von diesem System entfernen."

 LangString str_ungroup_plugin       ${LANG_GERMAN} \
-    "Entferne vimfiles Verzeichnis"
+    "Entferne Plugin-Verzeichnisse"
 LangString str_desc_rm_plugin       ${LANG_GERMAN} \
-    "Entfernt das vimfiles Verzeichnis, falls es leer ist."
+    "Entferne Plugin-Verzeichnisse, falls sie leer sind."

 LangString str_unsection_plugin_home ${LANG_GERMAN} \
     "Privat"
 LangString str_desc_rm_plugin_home  ${LANG_GERMAN} \
-    "Remove the plugin directories from HOME directory."
+    "Entfernt die Plugin-Verzeichnisse aus dem HOME Benutzerverzeichnis."

 LangString str_unsection_plugin_vim ${LANG_GERMAN} \
     "Freigegeben"
 LangString str_desc_rm_plugin_vim   ${LANG_GERMAN} \
-    "Remove the plugin directories from Vim install directory."
+    "Entfernt das Plugin-Verzeichnis aus dem Vim-Installationsverzeichnis."

 LangString str_unsection_rootdir    ${LANG_GERMAN} \
     "Entferne Vim Installationsverzeichnis"
```

k-takata commented:
> Why is `str_ungroup_plugin` and `str_desc_rm_plugin` in English plural? ("Remove plugin directories") 

Because `str_group_plugin` and `str_desc_plugin` are both plural. I think they mention $HOME/vimfiles and $VIM/vimfiles. (Or the directories under vimfiles (colors, compiler, doc, ...)? Not sure.)

> Where does it remove the plugin directories from?

It is a description for a section group which includes `str_unsection_plugin_home` and `str_unsection_plugin_vim`, so it removes from $HOME/vimfiles or $VIM/vimfiles.

k-takata committed:
nsis: Update German translation (by chrisbra)

https://github.com/vim/vim/pull/3501#issuecomment-444017314

scootergrisen committed:
Update danish.nsi

k-takata commented:
Thank you @chrisbra and @scootergrisen.
I think it's ready to be merged.

k-takata committed:
nsis: Remove outdated comments

Remove old comments from the translation files.
